lib/Controller/SMimeCertificatesController.php

```diff
@@ -0,0 +1,131 @@
+	private SMimeCertificateMapper $certificateMapper;
+
+	public function __construct(string $appName,
+								IRequest $request,
+								?string $userId,
+								SMimeService $certificateService,
+								SMimeCertificateMapper $certificateMapper) {
+		parent::__construct($appName, $request);
+		$this->userId = $userId;
+		$this->certificateService = $certificateService;
+		$this->certificateMapper = $certificateMapper;
+	}
+
+	/**
+	 * @NoAdminRequired
+	 * @TrapError
+	 *
+	 * @throws ServiceException
+	 */
+	public function index(): JSONResponse {
+		$certificates = $this->certificateMapper->findAll($this->userId);
+		foreach ($certificates as $certificate) {
+			$this->certificateService->decryptCertificateEntity($certificate);
+		}
+		return new JSONResponse($certificates);
+	}
+
+	/**
+	 * @NoAdminRequired
+	 * @TrapError
+	 *
+	 * @param int $id
+	 * @return JSONResponse
+	 *
+	 * @throws DoesNotExistException
+	 * @throws ServiceException
+	 */
+	public function destroy(int $id): JSONResponse {
+		$certificate = $this->certificateMapper->find($id, $this->userId);
+		$this->certificateMapper->delete($certificate);
+		return new JSONResponse();
+	}
+
+	/**
+	 * @NoAdminRequired
+	 * @TrapError
+	 *
+	 * @return JSONResponse
+	 *
+	 * @throws ServiceException
+	 */
+	public function create(): JSONResponse {
+		// TODO: What about PKCS12 certificates?
+		// They need to be decrypted by the client because they are protected by a password.
+		// We could use
+		//  - https://w3c.github.io/FileAPI/#reading-data-section
+		//  - https://github.com/digitalbazaar/forge#pkcs12
+		// to read and decrypt a PKCS12 certificate in the browser.
+		// The decrypted data could be attached to hidden file inputs and then sent to this endpoint
+		// as well.
+
+		$attachedCertificate = $this->request->getUploadedFile('certificate');
+		$attachedKey = $this->request->getUploadedFile('privateKey');
+
+		if ($attachedCertificate === null) {
+			throw new ServiceException('No certificate attached');
+		}
+
+		$certificateFile = new UploadedFile($attachedCertificate);
+		$certificateData = file_get_contents($certificateFile->getTempPath());
+		$emailAddress = $this->certificateService->extractEmailAddress($certificateData);
+
+		$entity = new SMimeCertificate();
+		$entity->setUserId($this->userId);
+		$entity->setEmailAddress($emailAddress);
+		$entity->setCertificate($certificateData);
+		if ($attachedKey !== null) {
+			$keyFile = new UploadedFile($attachedKey);
+			$privateKeyData = file_get_contents($keyFile->getTempPath());
+			if (!$this->certificateService->checkPrivateKey($certificateData, $privateKeyData)) {
+				throw new ServiceException('Private key does not match certificate or is protected by a passphrase');
+			}
+			$entity->setPrivateKey($privateKeyData);
+		}
+		$this->certificateService->encryptCertificateEntity($entity);
+		$entity = $this->certificateMapper->insert($entity);
+		$this->certificateService->decryptCertificateEntity($entity);
```

Comment:
this can be moved to the service layer except for the file upload handling. the latter is responsibility of the controller. the service will receive the key as nullable string.

Reply:
Done